Entities/SidewaysJumpThru.cs

```diff
@@ -0,0 +1,285 @@
+                if (next.OpCode == OpCodes.Call && (next.Operand as MethodReference)?.FullName == "System.Boolean Monocle.Entity::CollideCheck<Celeste.Solid>(Microsoft.Xna.Framework.Vector2)") {
+                    Logger.Log("SpringCollab2020/SidewaysJumpThru", $"Patching Entity.CollideCheck to include sideways jumpthrus at {cursor.Index} in IL for {il.Method.Name}");
+
+                    cursor.Remove();
+                    cursor.EmitDelegate<Func<Entity, Vector2, bool>>((self, checkAtPosition) => {
+                        // we still want to check for solids...
+                        if (self.CollideCheck<Solid>(checkAtPosition))
+                            return true;
+
+                        // if we are not checking a side, this certainly has nothing to do with jumpthrus.
+                        if (self.Position.X == checkAtPosition.X)
+                            return false;
+
+                        // our entity also collides if this is with a jumpthru and we are colliding with the solid side of it.
+                        // we are in this case if the jumpthru is left to right (the "solid" side of it is the right one) 
+                        // and we are checking the collision on the left side of the player for example.
+                        bool collideOnLeftSideOfPlayer = (self.Position.X > checkAtPosition.X);
+                        SidewaysJumpThru jumpthru = self.CollideFirstOutside<SidewaysJumpThru>(checkAtPosition);
+                        return jumpthru != null && self is Player player && (jumpthru.AllowLeftToRight == collideOnLeftSideOfPlayer);
+                    });
+                }
+
+                if (next.OpCode == OpCodes.Callvirt && (next.Operand as MethodReference)?.FullName == "System.Boolean Monocle.Scene::CollideCheck<Celeste.Solid>(Microsoft.Xna.Framework.Vector2)") {
+                    Logger.Log("SpringCollab2020/SidewaysJumpThru", $"Patching Scene.CollideCheck to include sideways jumpthrus at {cursor.Index} in IL for {il.Method.Name}");
+
+                    cursor.Remove();
+                    cursor.EmitDelegate<Func<Scene, Vector2, bool>>((self, vector) => self.CollideCheck<Solid>(vector) || self.CollideCheck<SidewaysJumpThru>(vector));
+                }
+
+                cursor.Index++;
+            }
+        }
+
+        private static int onPlayerNormalUpdate(On.Celeste.Player.orig_NormalUpdate orig, Player self) {
+            int result = orig(self);
+
+            // kill speed if player is going towards a jumpthru.
+            if (self.Speed.X != 0) {
+                bool movingLeftToRight = self.Speed.X > 0;
+                SidewaysJumpThru jumpThru = self.CollideFirstOutside<SidewaysJumpThru>(self.Position + Vector2.UnitX * Math.Sign(self.Speed.X));
+                if (jumpThru != null && jumpThru.AllowLeftToRight != movingLeftToRight) {
+                    self.Speed.X = 0;
+                }
+            }
+
+            return result;
+        }
+
+
+

```

Comment:
What's up with the whitespace here?

Reply:
That was to separate hook code from the entity code 😅 I think I should drop a comment in here instead.